src/java.base/share/classes/java/io/Reader.java

```diff
@@ -457,7 +449,73 @@ private String readAllCharsAsString() throws IOException {
      * @since 25
      */
     public List<String> readAllLines() throws IOException {
-        return readAllCharsAsString().lines().toList();
+        List<String> lines = new ArrayList<>();
+
+        int writePos = 0;
+        int fragPos = -1;
+        int fragLen = 0;
+
+        char[] cb = new char[TRANSFER_BUFFER_SIZE];
+        boolean skipLF = false;
+        int n;
+        while ((n = read(cb, writePos, cb.length - writePos)) != -1) {
+            int pos = writePos;
+            int limit = pos + n;
+            while (pos < limit) {
+                // find next line terminator
+                int term = pos;
+                while (term < limit) {
+                    char c = cb[term];
+                    if (c == '\n' || c == '\r')
+                        break;
+                    term++;
+                }
+
+                if (term < limit) { // line terminator
+                    boolean isCR = (cb[term] == '\r');
+                    if (isCR || !(skipLF && term == pos)) {
+                        // line terminator is a CR or an LF just after a CR
+                        if (fragPos != -1) {
+                            lines.add(new String(cb, fragPos, term - fragPos));
+                            fragPos = -1;
+                        } else {
+                            lines.add(new String(cb, pos, term - pos));
+                        }
+                    }
```

Comment:
This might be a good place to check for NL after CR and efficiently skip it.  (Common on Windows).
It could save a cycle through the `while (pos < limit)` loop just to ignore the LF. (Except for the case of CR at the end of the buffer.)

Reply:
Fixed in c9d1ab3.